cloudigrade/account/reports/__init__.py

```diff
@@ -179,3 +187,113 @@ def _calculate_instance_usage(start, end, events):
 
     return time_running
+
+def validate_event(event, start):
+    """
+    Ensure that the event is relevant to our time frame.
+
+    Args:
+        event: (InstanceEvent): The event object to evaluate
+        start (datetime.datetime): Start time (inclusive)
+
+    Returns:
+        bool: A boolean regarding whether or not we should inspect the event further.
+    """
+    valid_event = True
+    # if the event occurred outside of our specified period (ie. before start) we should
+    # only inspect it if it was a power on event
+    if event.occurred_at < start:
+        if event.event_type == InstanceEvent.TYPE.power_off:
+            valid_event = False
+    return valid_event
+
+def validate_account_creation(account, end):
+    """
+    Ensure that the event is relevant to our time frame.
+
+    Args:
+        account: (Account): The event object to evaluate
+        end (datetime.datetime): End time (exclusive)
+
+    Returns:
+        bool: A boolean regarding whether or not the account was created after the end
+        time.
+    """
+    if end <= account.created_at:
+        return False
+    return True
+
+
+def get_account_overview(account, start, end):
+    """
+    Generate an overview of an account over a specified amount of time.
+
+    Args:
+        start (datetime.datetime): Start time (inclusive)
+        end (datetime.datetime): End time (exclusive)
+        account (Account or AwsAccount): Account or AwsAccount object
+
+    Returns:
+        dict: An overview of the instances/images/rhel & openshift images for the specified
+        account during the specified time period."""
+    machine_images = AwsMachineImage.objects.all()
+    instances = []
+    images = []
+    rhel = []
+    openshift = []
+    cloud_type = 'unknown'
+    arn = 'unsupported'
+    # make sure that this is a supported cloud type (aws at this time)
+    if isinstance(account, AwsAccount):
+        cloud_helper = helper.get_report_helper(AWS_PROVIDER_STRING, account.aws_account_id)
+        # _get_relevant_events will return the events in between the start & end times & if
+        # no events are present during this period, it will return the last event that occurred
+        events = _get_relevant_events(start, end, cloud_helper)
+        for event in events:
+            valid_event = validate_event(event, start)
+            if valid_event:
+                instances.append(event.instance.id)
+                images.append(event.ec2_ami_id)
+                image = machine_images.filter(ec2_ami_id=event.ec2_ami_id).first()
+                if image:
+                    for tag in image.tags.all():
+                        if tag.description == 'rhel':
+                            rhel.append(image)
+                        if tag.description == 'openshift':
+                            openshift.append(image)
+        # set aws account info & grab the totals
+        cloud_type = AWS_PROVIDER_STRING
```

Comment:
I suspect this isn't (or won't be) the only place we would directly reference `AWS_PROVIDER_STRING` in places we're dealing specifically with `AwsAccount` objects. Wouldn't it be nice if `Account` objects had a property that we could just look at instead of inferring based on things like the `isinstance(account, AwsAccount)` a few lines above this?

Now might be a good time to add that. If so, I would add an abstract property to the `Account` class and an actual property to the `AwsAccount` class that returns `AWS_PROVIDER_STRING` directly. Then, whenever we need to know the provider, we can just look at the instance like `account.cloud_provider`, and that's a lot cleaner than `isinstance` checking. If you're not sure how to do this, see `product_identifier` as a similar implementation on `InstanceEvent`/`AwsInstanceEvent`.

...or maybe my suggestion is overkill? I took a look, and there currently aren't many places `AWS_PROVIDER_STRING` is actually referenced, but we _will_ have to do something _like_ this when we expand beyond AWS. Food for thought.

Reply:
I definitely agree that it would be nice to have the type property built into the accounts and we should plan on this for the future, but I don't think that we should add onto this issue to do it.